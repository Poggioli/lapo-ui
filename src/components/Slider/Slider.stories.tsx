import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';
import { Slider } from './Slider';

export default {
  title: 'Componentes/Slider',
  component: Slider,
  argTypes: {
    min: {
      type: 'number',
      defaultValue: 0
    },
    max: {
      type: 'number',
      defaultValue: 100
    },
    orientation: {
      control: 'radio',
      options: ['horizontal', 'vertical'],
      defaultValue: 'horizontal'
    },
    step: {
      type: 'number',
      defaultValue: 1
    },
    minStepsBetweenThumbs: {
      type: 'number',
      defaultValue: 10
    },
    showRange: {
      type: 'boolean',
      defaultValue: false
    },
    inverted: {
      type: 'boolean',
      defaultValue: false
    },
    disabled: {
      type: 'boolean',
      defaultValue: false
    },
    ref: {
      table: {
        disable: true
      }
    },
    css: {
      table: {
        disable: true
      }
    },
    as: {
      table: {
        disable: true
      }
    }
  }
} as ComponentMeta<typeof Slider>;

const Template: ComponentStory<typeof Slider> = ({ ...args }) => <Slider {...args} />;

export const Default = Template.bind({});
