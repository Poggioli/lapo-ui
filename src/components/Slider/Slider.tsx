import { ComponentCSSProp, LapoVariants } from '@lapo';
import type * as Polymorphic from '@radix-ui/react-polymorphic';
import useId from '@utils/hooks/useId';
import React, { useCallback, useMemo, useState } from 'react';
import {
  Container,
  RangeValue,
  StyledRange,
  StyledRoot,
  StyledThumb,
  StyledTrack,
  StyledValue
} from './styles';

/* -------------------------------------------------------------------------------------------------
 * Slider
 * -----------------------------------------------------------------------------------------------*/

type SliderCSSProp = ComponentCSSProp;
type SliderVariants = LapoVariants<typeof StyledRoot>;
type SliderProps = SliderCSSProp &
  SliderVariants & {
    showRange?: boolean;
  };

type SliderComponent = Polymorphic.ForwardRefComponent<typeof StyledRoot, SliderProps>;

const Slider = React.forwardRef(
  (
    {
      id,
      showRange,
      min,
      max,
      inverted,
      defaultValue,
      orientation,
      value,
      disabled,
      onValueChange,
      ...props
    },
    forwardedRef
  ) => {
    const ownId: string = useMemo(() => id || useId('slider'), [id]);
    const ownDefaultValue: number[] = defaultValue || value || [0];
    const ownMin: number = min || 0;
    const ownMax: number = max || 100;
    const [ownValue, setownValue] = useState<number[]>(value || ownDefaultValue);

    const ownOnValueChange = useCallback((value: number[]) => {
      onValueChange && onValueChange(value);
      setownValue(value);
    }, []);

    return (
      <Container orientation={orientation}>
        {showRange ? (
          <RangeValue
            role="range"
            aria-label={`${inverted ? ownMax : ownMin}`}
            variant="body1"
            disabled={disabled}
          >
            {inverted ? ownMax : ownMin}
          </RangeValue>
        ) : null}
        <StyledRoot
          {...props}
          id={ownId}
          ref={forwardedRef}
          disabled={disabled}
          inverted={inverted}
          min={ownMin}
          max={ownMax}
          orientation={orientation}
          defaultValue={ownDefaultValue}
          onValueChange={ownOnValueChange}
        >
          <StyledTrack>
            <StyledRange />
          </StyledTrack>
          {ownDefaultValue.map((_, index) => (
            <StyledThumb key={index} aria-label={`Slider ${index}`}>
              <StyledValue variant="body2">{ownValue[index]}</StyledValue>
            </StyledThumb>
          ))}
        </StyledRoot>
        {showRange ? (
          <RangeValue
            role="range"
            aria-label={`${inverted ? ownMin : ownMax}`}
            variant="body1"
            disabled={disabled}
          >
            {inverted ? ownMin : ownMax}
          </RangeValue>
        ) : null}
      </Container>
    );
  }
) as SliderComponent;

export { Slider };

/* -----------------------------------------------------------------------------------------------*/
