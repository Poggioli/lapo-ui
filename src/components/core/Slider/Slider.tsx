import { ComponentCSSProp, LapoVariants } from '@lapo';
import type * as Polymorphic from '@radix-ui/react-polymorphic';
import useId from '@utils/hooks/useId';
import React, { useCallback, useMemo, useState } from 'react';
import {
  Container,
  RangeValue,
  StyledRange,
  StyledRoot,
  StyledThumb,
  StyledTrack,
  StyledValue
} from './styles';

/* -------------------------------------------------------------------------------------------------
 * Slider
 * -----------------------------------------------------------------------------------------------*/

type SliderCSSProp = ComponentCSSProp;
type SliderVariants = LapoVariants<typeof StyledRoot>;
type SliderProps = SliderCSSProp &
  SliderVariants & {
    showRange?: boolean;
  };

type SliderComponent = Polymorphic.ForwardRefComponent<typeof StyledRoot, SliderProps>;

const Slider = React.forwardRef(
  (
    {
      id,
      showRange,
      min,
      max,
      inverted,
      defaultValue,
      orientation,
      value,
      disabled,
      onValueChange,
      ...props
    },
    forwardedRef
  ) => {
    const ownId: string = useMemo(() => id || useId('slider'), [id]);
    const ownDefaultValue: number[] = defaultValue || value || [0];
    const [ownValue, setOwnValue] = useState<number[]>(value || ownDefaultValue);

    const ownOnValueChange = useCallback((value: number[]) => {
      onValueChange && onValueChange(value);
      setOwnValue(value);
    }, []);

    return (
      <Container orientation={orientation}>
        {showRange ? (
          <RangeValue
            role="range"
            aria-label={`${inverted ? max : min}`}
            variant="body1"
            disabled={disabled}
          >
            {inverted ? max : min}
          </RangeValue>
        ) : null}
        <StyledRoot
          {...props}
          id={ownId}
          ref={forwardedRef}
          disabled={disabled}
          inverted={inverted}
          min={min}
          max={max}
          orientation={orientation}
          defaultValue={ownDefaultValue}
          onValueChange={ownOnValueChange}
        >
          <StyledTrack>
            <StyledRange />
          </StyledTrack>
          {ownDefaultValue.map((_, index) => (
            <StyledThumb key={index} aria-label={`Slider ${index}`}>
              <StyledValue variant="body2">{ownValue[index]}</StyledValue>
            </StyledThumb>
          ))}
        </StyledRoot>
        {showRange ? (
          <RangeValue
            role="range"
            aria-label={`${inverted ? min : max}`}
            variant="body1"
            disabled={disabled}
          >
            {inverted ? min : max}
          </RangeValue>
        ) : null}
      </Container>
    );
  }
) as SliderComponent;

Slider.defaultProps = {
  showRange: false,
  disabled: false,
  inverted: false,
  min: 0,
  max: 100,
  orientation: 'horizontal'
};

export { Slider };

/* -----------------------------------------------------------------------------------------------*/
