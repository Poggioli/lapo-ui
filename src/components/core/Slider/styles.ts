import { StyledText } from '@components/typograph/Text';
import { styled } from '@lapo';
import { Flex } from '@components/layout/Flex';
import * as SliderPrimitive from '@radix-ui/react-slider';

const Container = styled(Flex, {
  minHeight: '$9',
  gap: '$4',
  width: 'fit-content',

  variants: {
    orientation: {
      horizontal: {
        alignItems: 'flex-end !important',
        flexDirection: 'row !important'
      },
      vertical: {
        alignItems: 'center !important',
        flexDirection: 'column !important'
      }
    }
  },

  defaultVariants: {
    orientation: 'horizontal'
  }
});

const RangeValue = styled(StyledText, {
  color: '$slider-color-max-and-min-value-normal !important',
  transition: 'color 200ms linear',

  variants: {
    disabled: {
      true: {
        color: '$slider-color-max-and-min-value-disabled !important'
      }
    }
  },

  defaultVariants: {
    disabled: false
  }
});

const StyledValue = styled(StyledText, {
  height: 0,
  opacity: 0,
  pointerEvents: 'none',
  position: 'absolute',
  transition: `color 200ms linear,
    height 200ms linear,
    left 300ms linear,
    opacity 300ms linear,
    top 300ms linear,
    width 200ms linear`,
  width: 0,
  color: '$slider-color-indicator-value-normal !important'
});
StyledValue.toString = () => `.${StyledValue.className}`;

const StyledRange = styled(SliderPrimitive.Range, {
  position: 'absolute',
  backgroundColor: '$slider-background-color-range-normal',
  borderRadius: '$4',
  transition: 'background-color 200ms linear',

  '&[data-disabled]': {
    backgroundColor: '$slider-background-color-range-disabled'
  }
});
StyledRange.toString = () => `.${StyledRange.className}`;

const StyledTrack = styled(SliderPrimitive.Track, {
  backgroundColor: '$slider-background-color-track-normal',
  borderRadius: '$4',
  flexGrow: 1,
  position: 'relative',
  transition: 'background-color 200ms linear',

  '&[data-disabled]': {
    backgroundColor: '$slider-background-color-track-disabled'
  },

  '&[data-orientation="horizontal"]': {
    height: '$2',

    [`${StyledRange.toString()}`]: {
      height: '100%'
    }
  },

  '&[data-orientation="vertical"]': {
    width: '$2',

    [`${StyledRange.toString()}`]: {
      width: '100%'
    }
  }
});

const StyledThumb = styled(SliderPrimitive.Thumb, {
  all: 'unset',
  backgroundColor: '$slider-background-color-thumb-normal',
  borderRadius: '$3',
  boxShadow: '$$boxShadow',
  display: 'block',
  height: '$4',
  transition: `background-color 200ms linear,
    transform 200ms linear,
    box-shadow 200ms linear`,
  width: '$4',
  position: 'relative',

  $$boxShadowColor: '$colors-slider-box-shadow-thumb-normal',
  $$boxShadow: '0 0 0 $$boxShadowColor',

  '&[data-disabled]': {
    backgroundColor: '$slider-background-color-thumb-disabled'
  },

  '&:not([data-disabled])': {
    '&:hover': {
      transform: 'scale(1.25)',
      $$boxShadowColor: '$colors-slider-box-shadow-thumb-hover',
      $$boxShadow: '0 2px 10px $$boxShadowColor',
      backgroundColor: '$slider-background-color-thumb-hover'
    },

    '&:focus': {
      transform: 'scale(1.25)',
      $$boxShadowColor: '$colors-slider-box-shadow-thumb-focus',
      $$boxShadow: '0 2px 10px $$boxShadowColor',
      backgroundColor: '$slider-background-color-thumb-focus'
    },

    '&:active': {
      transform: 'scale(1.25)',
      $$boxShadowColor: '$colors-slider-box-shadow-thumb-active',
      $$boxShadow: '0 2px 10px $$boxShadowColor',
      backgroundColor: '$slider-background-color-thumb-active'
    }
  }
});
StyledThumb.toString = () => `.${StyledThumb.className}`;

const StyledRoot = styled(SliderPrimitive.Root, {
  position: 'relative',
  display: 'flex',
  alignItems: 'center',
  userSelect: 'none',
  touchAction: 'none',

  '&[data-orientation="horizontal"]': {
    height: '$5',
    minWidth: 'calc($13 * 2)',
    width: '100%',

    '&:not([data-disabled])': {
      [`${StyledThumb.toString()} ${StyledValue.toString()}`]: {
        top: 0
      },

      [`${StyledThumb.toString()}:is(:hover, :focus, :active) ${StyledValue.toString()}`]: {
        height: '100%',
        opacity: 1,
        top: '-$5',
        width: '100%'
      }
    }
  },

  '&[data-orientation="vertical"]': {
    flexDirection: 'column',
    height: '100%',
    minHeight: 'calc($13 * 2)',
    width: '$5',

    '&:not([data-disabled])': {
      [`${StyledThumb.toString()} ${StyledValue.toString()}`]: {
        left: 0
      },

      [`${StyledThumb.toString()}:is(:hover, :focus, :active) ${StyledValue.toString()}`]: {
        height: '100%',
        opacity: 1,
        left: '$5',
        width: '100%'
      }
    }
  }
});

export { RangeValue, StyledRoot, StyledTrack, StyledRange, StyledThumb, StyledValue, Container };
