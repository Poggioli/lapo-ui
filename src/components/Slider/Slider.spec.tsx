import { LapoVariants } from '@lapo';
import { render, RenderResult } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';
import React, { FC, useState } from 'react';
import { Slider } from './Slider';

describe('Given a <Slider /> component', () => {
  let rendered: RenderResult;

  const mount = ({ ...props }: LapoVariants<typeof Slider>) => {
    rendered = render(<Slider {...props} />);
  };

  it('should have no a11y violations', async () => {
    mount({});
    expect(await axe(rendered.container)).toHaveNoViolations();
  });

  describe('value and defaultValue', () => {
    it('should set defaultValue when pass a defaultValue', () => {
      mount({
        defaultValue: [50]
      });

      const slider = rendered.getByRole('slider');
      expect(slider).toHaveAttribute('aria-valuenow', '50');
    });

    it('should set defaultValue when pass a value', () => {
      mount({
        value: [50, 80]
      });

      const slider = rendered.getAllByRole('slider');
      expect(slider[0]).toHaveAttribute('aria-valuenow', '50');
      expect(slider[1]).toHaveAttribute('aria-valuenow', '80');
    });

    it('should set defaultValue as [0] when dont pass any defaultValue and any value', () => {
      mount({});

      const slider = rendered.getByRole('slider');
      expect(slider).toHaveAttribute('aria-valuenow', '0');
    });
  });

  describe('Should have id', () => {
    it('With prefix when no Id is provided', () => {
      mount({});
      const slider = rendered.container.querySelector('[id^="slider-"]');
      expect(slider).toBeVisible();
    });

    it('With Id passed by props', () => {
      mount({ id: 'sliderId123' });
      const slider = rendered.container.querySelector('#sliderId123');
      expect(slider).toBeVisible();
    });
  });

  describe('when pass value to showRange', () => {
    it('should show if its true', () => {
      const minValue = 1000;
      const maxValue = 5000;
      mount({
        showRange: true,
        min: minValue,
        max: maxValue
      });

      const minValueLabel = rendered.getByRole('range', {
        name: `${minValue}`
      });
      const maxValueLabel = rendered.getByRole('range', {
        name: `${maxValue}`
      });

      expect(minValueLabel).toBeVisible();
      expect(maxValueLabel).toBeVisible();
    });

    it('should not show if its false', () => {
      mount({});

      const minValueLabel = rendered.queryByRole('range');
      const maxValueLabel = rendered.queryByRole('range');

      expect(minValueLabel).not.toBeInTheDocument();
      expect(maxValueLabel).not.toBeInTheDocument();
    });
  });

  it('should invert range label if inverted value is true', () => {
    mount({
      inverted: true,
      id: 'sliderId123'
    });

    expect(rendered).toMatchSnapshot();
  });

  // ToDo try to test the onValueChange
  xit('should call onValueChange when it`s not undefined', () => {
    const onValueChangeMock = jest.fn();
    mount({
      onValueChange: onValueChangeMock,
      id: 'sliderId123'
    });

    const slider = rendered.container.querySelector('#sliderId123');
    userEvent.click(slider);

    expect(onValueChangeMock).toHaveBeenCalled();
  });
});
