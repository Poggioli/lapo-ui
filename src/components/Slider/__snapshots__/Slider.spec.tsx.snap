// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a <Slider /> component should invert range label if inverted value is true 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="lapo--c-lesPJm lapo--c-dhzjXW lapo--c-lesPJm-igAUidj-css lapo--c-jJvGnk lapo--c-jJvGnk-dJfSKz-orientation-horizontal"
      >
        <span
          aria-disabled="false"
          class="lapo--c-gVJsmb"
          data-orientation="horizontal"
          dir="ltr"
          id="sliderId123"
          style="--radix-slider-thumb-transform: translateX(-50%);"
        >
          <span
            class="lapo--c-EMfmZ"
            data-orientation="horizontal"
          >
            <span
              class="lapo--c-fQqFYj"
              data-orientation="horizontal"
              style="right: 0%; left: 100%;"
            />
          </span>
          <span
            style="transform: var(--radix-slider-thumb-transform); position: absolute; right: calc(0% + 0px);"
          >
            <span
              aria-label="Slider 0"
              aria-orientation="horizontal"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="0"
              class="lapo--c-lcUwFu"
              data-orientation="horizontal"
              data-radix-collection-item=""
              role="slider"
              style=""
              tabindex="0"
            >
              <span
                class="lapo--c-jNLug lapo--c-cLMHMJ lapo--c-jNLug-bkzKcy-variant-body2 lapo--c-jNLug-ewLzLC-color-grayColoredHighContrast"
              >
                0
              </span>
            </span>
          </span>
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="lapo--c-lesPJm lapo--c-dhzjXW lapo--c-lesPJm-igAUidj-css lapo--c-jJvGnk lapo--c-jJvGnk-dJfSKz-orientation-horizontal"
    >
      <span
        aria-disabled="false"
        class="lapo--c-gVJsmb"
        data-orientation="horizontal"
        dir="ltr"
        id="sliderId123"
        style="--radix-slider-thumb-transform: translateX(-50%);"
      >
        <span
          class="lapo--c-EMfmZ"
          data-orientation="horizontal"
        >
          <span
            class="lapo--c-fQqFYj"
            data-orientation="horizontal"
            style="right: 0%; left: 100%;"
          />
        </span>
        <span
          style="transform: var(--radix-slider-thumb-transform); position: absolute; right: calc(0% + 0px);"
        >
          <span
            aria-label="Slider 0"
            aria-orientation="horizontal"
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuenow="0"
            class="lapo--c-lcUwFu"
            data-orientation="horizontal"
            data-radix-collection-item=""
            role="slider"
            style=""
            tabindex="0"
          >
            <span
              class="lapo--c-jNLug lapo--c-cLMHMJ lapo--c-jNLug-bkzKcy-variant-body2 lapo--c-jNLug-ewLzLC-color-grayColoredHighContrast"
            >
              0
            </span>
          </span>
        </span>
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
